giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+          }
+          if (extreme[4] != -1) {
+            coefficient[4] = extreme[4];
+          }
+          setIsValid = true;
+          return; // the finally-block will execute before return
+        }
+        // CHECKSTYLE: stop IllegalCatch
+      } catch (Exception e) {
+        // CHECKSTYLE: resume IllegalCatch
+        LOG.warn("addRecord: exception occurred!", e);
+      } finally {
+        // This inner try-finally block is necessary to ensure that the
+        // lock is always released.
+        try {
+          isValid = setIsValid;
+          printStats();
+        } finally {
+          lock.unlock();
+        }
+      }
+    }
+
+    /**
+     * Certain coefficients need to be within a specific range. For instance,
+     * If the coefficient is not in this range, we set it to the closest bound
+     * and re-run the linear regression.
+     *
+     * @param coefIndex Coefficient index
+     * @param lowerBound Lower bound
+     * @param upperBound Upper bound
+     * @param xValues double[][] matrix with data samples
+     * @param yValues double[] matrix with y samples
+     * @return True if coefficients were out-of-range
+     * @throws Exception
+     */
+    private boolean refineCoefficient(int coefIndex, double lowerBound,
+      double upperBound, double[][] xValues, double[] yValues)
+      throws Exception {
+
+      boolean result = false;
+      if (coefficient[coefIndex] < lowerBound ||
+        coefficient[coefIndex] > upperBound) {
+
+        double value;
+        if (coefficient[coefIndex] < lowerBound) {
+          value = lowerBound;
+        } else {
+          value = upperBound;
+        }
+        int ptr = -1;
+        if (validColumnIndices.size() >= 1 &&
+          validColumnIndices.get(validColumnIndices.size() - 1) == coefIndex) {
+          ptr = validColumnIndices.size() - 1;
+        } else if (validColumnIndices.size() >= 2 &&
+          validColumnIndices.get(validColumnIndices.size() - 2) == coefIndex) {
+          ptr = validColumnIndices.size() - 2;
+        }
+        if (ptr != -1) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("addRecord: coefficient at index " + coefIndex +
+              " is wrong in the regression, setting it to " + value);
+          }
+          // remove from valid column
+          validColumnIndices.remove(ptr);
+          // re-create the X matrix
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          // adjust Y values
+          for (int i = 0; i < memorySamples.size(); ++i) {
+            yValues[i] -= value * dataSamples.get(i)[coefIndex];
+          }
+          // save new coefficient value in intermediate array
+          extreme[coefIndex] = value;
+          // re-run regression
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+          result = true;
+        } else {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
```

Comment:
I'll add more comment for the `refineCoefficient` method to clarify what it does.